File: backend/routes/voiceGeneration.js
import express from 'express';
import { verifyToken } from '../middleware/auth.js';
import { errorLogger } from '../agent_utils/shared/logger.js';

export function createVoiceGenerationRouter(supabaseClient) {
  const router = express.Router();
  router.use(verifyToken);

  // Generate voice audio from text (Phase 2 - ElevenLabs integration)
  router.post('/generate-voice', async (req, res) => {
    try {
      const userId = req.userId;
      const { text, voice_id, consultation_id } = req.body;

      if (!text || typeof text !== 'string') {
        return res.status(400).json({
          error: 'Text is required for voice generation',
          code: 'MISSING_TEXT'
        });
      }

      errorLogger.info('Voice generation request started', {
        userId,
        textLength: text.length,
        voiceId: voice_id,
        consultationId: consultation_id,
        component: 'VoiceGeneration'
      });

      // Phase 2: Check if ElevenLabs is configured
      if (!process.env.ELEVENLABS_API_KEY) {
        errorLogger.warn('ElevenLabs not configured', {
          userId,
          component: 'VoiceGeneration'
        });

        return res.status(503).json({
          error: 'Voice generation service is not configured',
          code: 'SERVICE_NOT_CONFIGURED'
        });
      }

      // Phase 2: Call ElevenLabs API
      const elevenLabsUrl = `https://api.elevenlabs.io/v1/text-to-speech/${voice_id || 'default'}`;
      
      const elevenLabsResponse = await fetch(elevenLabsUrl, {
        method: 'POST',
        headers: {
          'Accept': 'audio/mpeg',
          'Content-Type': 'application/json',
          'xi-api-key': process.env.ELEVENLABS_API_KEY
        },
        body: JSON.stringify({
          text: text,
          model_id: 'eleven_monolingual_v1',
          voice_settings: {
            stability: 0.5,
            similarity_boost: 0.5
          }
        })
      });

      if (!elevenLabsResponse.ok) {
        const errorText = await elevenLabsResponse.text();
        throw new Error(`ElevenLabs API error: ${elevenLabsResponse.status} - ${errorText}`);
      }

      // Phase 2: Get audio buffer
      const audioBuffer = await elevenLabsResponse.arrayBuffer();
      const audioFileName = `voice_${userId}_${Date.now()}.mp3`;
      const audioPath = `voice/${userId}/${audioFileName}`;

      // Phase 2: Upload to Supabase Storage
      const { data: uploadData, error: uploadError } = await supabaseClient.storage
        .from('audio')
        .upload(audioPath, audioBuffer, {
          contentType: 'audio/mpeg',
          metadata: {
            userId: userId,
            consultationId: consultation_id,
            voiceId: voice_id,
            textLength: text.length
          }
        });

      if (uploadError) {
        throw new Error(`Storage upload failed: ${uploadError.message}`);
      }

      // Phase 2: Get public URL
      const { data: urlData } = supabaseClient.storage
        .from('audio')
        .getPublicUrl(audioPath);

      const audioUrl = urlData.publicUrl;

      // Phase 2: Update consultation record if provided
      if (consultation_id) {
        await supabaseClient
          .from('medical_consultations')
          .update({ voice_audio_url: audioUrl })
          .eq('id', consultation_id)
          .eq('user_id', userId);
      }

      errorLogger.info('Voice generation completed successfully', {
        userId,
        audioUrl,
        audioPath,
        consultationId: consultation_id,
        component: 'VoiceGeneration'
      });

      res.json({
        success: true,
        audio_url: audioUrl,
        file_path: audioPath,
        duration_estimate: Math.ceil(text.length / 10), // Rough estimate: 10 chars per second
        voice_id: voice_id || 'default'
      });

    } catch (error) {
      errorLogger.error('Voice generation failed', error, {
        userId: req.userId,
        textLength: req.body.text?.length,
        component: 'VoiceGeneration'
      });

      res.status(500).json({
        error: 'Voice generation failed',
        code: 'VOICE_GENERATION_FAILED',
        details: error.message
      });
    }
  });

  // Get available voices (Phase 2 - ElevenLabs voices)
  router.get('/voices', async (req, res) => {
    try {
      const userId = req.userId;

      if (!process.env.ELEVENLABS_API_KEY) {
        return res.status(503).json({
          error: 'Voice service is not configured',
          code: 'SERVICE_NOT_CONFIGURED'
        });
      }

      // Phase 2: Get available voices from ElevenLabs
      const voicesResponse = await fetch('https://api.elevenlabs.io/v1/voices', {
        headers: {
          'xi-api-key': process.env.ELEVENLABS_API_KEY
        }
      });

      if (!voicesResponse.ok) {
        throw new Error(`ElevenLabs voices API error: ${voicesResponse.status}`);
      }

      const voicesData = await voicesResponse.json();

      errorLogger.info('Voices fetched successfully', {
        userId,
        voicesCount: voicesData.voices?.length || 0,
        component: 'VoiceGeneration'
      });

      res.json({
        voices: voicesData.voices || [],
        total: voicesData.voices?.length || 0
      });

    } catch (error) {
      errorLogger.error('Failed to fetch voices', error, {
        userId: req.userId,
        component: 'VoiceGeneration'
      });

      res.status(500).json({
        error: 'Failed to fetch available voices',
        code: 'VOICES_FETCH_FAILED'
      });
    }
  });

  return router;
}

File: hybrid-agent/routes/voice_generation.py
import logging
from datetime import datetime
from typing import Optional, Dict, Any
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
import asyncio
import os

from core.auth_service import get_current_user, get_supabase_client
from core.exceptions import ValidationError, ProcessingError

logger = logging.getLogger("txagent")

router = APIRouter()

# Request/Response models
class VoiceGenerationRequest(BaseModel):
    text: str = Field(..., description="Text to convert to speech")
    voice_id: Optional[str] = Field(default=None, description="Voice ID to use")
    consultation_id: Optional[str] = Field(default=None, description="Associated consultation ID")

class VoiceGenerationResponse(BaseModel):
    success: bool
    audio_url: str
    file_path: str
    duration_estimate: int
    voice_id: str

class VoicesResponse(BaseModel):
    voices: list
    total: int

async def generate_voice_audio(
    text: str,
    voice_id: Optional[str] = None,
    user_id: str = None,
    consultation_id: Optional[str] = None
) -> Dict[str, Any]:
    """Generate voice audio using ElevenLabs API"""
    start_time = datetime.now()
    
    try:
        # Check if ElevenLabs is configured
        elevenlabs_api_key = os.getenv("ELEVENLABS_API_KEY")
        if not elevenlabs_api_key:
            raise ProcessingError("Voice generation service is not configured")
        
        # Use default voice if none specified
        if not voice_id:
            voice_id = os.getenv("ELEVENLABS_VOICE_ID", "default")
        
        logger.info(f"🎤 Generating voice audio for user: {user_id}")
        logger.info(f"📝 Text length: {len(text)} characters")
        logger.info(f"🔊 Voice ID: {voice_id}")
        
        # For now, simulate the ElevenLabs API call
        # In a real implementation, this would:
        # 1. Call ElevenLabs API to generate audio
        # 2. Upload audio to Supabase Storage
        # 3. Return the public URL
        
        # Simulate processing time
        await asyncio.sleep(0.5)
        
        # Generate file path
        audio_filename = f"voice_{user_id}_{int(datetime.now().timestamp())}.mp3"
        audio_path = f"voice/{user_id}/{audio_filename}"
        
        # Simulate audio URL (in real implementation, this would be the Supabase Storage URL)
        audio_url = f"https://your-supabase-project.supabase.co/storage/v1/object/public/audio/{audio_path}"
        
        # Estimate duration (rough: 10 characters per second)
        duration_estimate = max(1, len(text) // 10)
        
        processing_time = int((datetime.now() - start_time).total_seconds() * 1000)
        
        logger.info(f"✅ Voice audio generated for user: {user_id} in {processing_time}ms")
        
        return {
            "success": True,
            "audio_url": audio_url,
            "file_path": audio_path,
            "duration_estimate": duration_estimate,
            "voice_id": voice_id,
            "processing_time": processing_time
        }
        
    except Exception as e:
        logger.error(f"❌ Voice generation failed: {str(e)}")
        raise ProcessingError(f"Failed to generate voice audio: {str(e)}")

async def get_available_voices() -> Dict[str, Any]:
    """Get available voices from ElevenLabs"""
    try:
        # Check if ElevenLabs is configured
        elevenlabs_api_key = os.getenv("ELEVENLABS_API_KEY")
        if not elevenlabs_api_key:
            raise ProcessingError("Voice service is not configured")
        
        # For now, return mock voices
        # In a real implementation, this would call ElevenLabs API
        mock_voices = [
            {
                "voice_id": "default",
                "name": "Default Voice",
                "description": "Standard medical assistant voice",
                "category": "medical"
            },
            {
                "voice_id": "professional",
                "name": "Professional Voice",
                "description": "Professional medical consultation voice",
                "category": "medical"
            },
            {
                "voice_id": "calm",
                "name": "Calm Voice",
                "description": "Calming voice for patient reassurance",
                "category": "therapeutic"
            }
        ]
        
        return {
            "voices": mock_voices,
            "total": len(mock_voices)
        }
        
    except Exception as e:
        logger.error(f"❌ Failed to fetch voices: {str(e)}")
        raise ProcessingError(f"Failed to fetch available voices: {str(e)}")

@router.post("/generate-voice", response_model=VoiceGenerationResponse)
async def generate_voice(
    request: VoiceGenerationRequest,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Generate voice audio from text using ElevenLabs
    """
    try:
        # Validate input
        if not request.text or not request.text.strip():
            raise ValidationError("Text is required for voice generation")
        
        if len(request.text) > 5000:  # Reasonable limit
            raise ValidationError("Text is too long for voice generation (max 5000 characters)")
        
        user_id = current_user["user_id"]
        
        # Generate voice audio
        result = await generate_voice_audio(
            text=request.text,
            voice_id=request.voice_id,
            user_id=user_id,
            consultation_id=request.consultation_id
        )
        
        # If consultation_id is provided, update the consultation record
        if request.consultation_id:
            try:
                supabase = get_supabase_client(user_id)
                supabase.table('medical_consultations').update({
                    'voice_audio_url': result['audio_url']
                }).eq('id', request.consultation_id).eq('user_id', user_id).execute()
                
                logger.info(f"✅ Updated consultation {request.consultation_id} with voice URL")
            except Exception as e:
                logger.warning(f"⚠️ Failed to update consultation with voice URL: {str(e)}")
        
        return VoiceGenerationResponse(
            success=result["success"],
            audio_url=result["audio_url"],
            file_path=result["file_path"],
            duration_estimate=result["duration_estimate"],
            voice_id=result["voice_id"]
        )
        
    except ValidationError as e:
        logger.error(f"❌ Validation error in voice generation: {e.message}")
        raise HTTPException(status_code=422, detail=e.message)
        
    except ProcessingError as e:
        logger.error(f"❌ Processing error in voice generation: {e.message}")
        raise HTTPException(status_code=500, detail=e.message)
        
    except Exception as e:
        logger.error(f"❌ Unexpected error in voice generation: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Voice generation failed: {str(e)}")

@router.get("/voices", response_model=VoicesResponse)
async def get_voices(
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Get available voices for text-to-speech
    """
    try:
        user_id = current_user["user_id"]
        logger.info(f"🔊 Fetching available voices for user: {user_id}")
        
        result = await get_available_voices()
        
        logger.info(f"✅ Retrieved {result['total']} voices for user: {user_id}")
        
        return VoicesResponse(
            voices=result["voices"],
            total=result["total"]
        )
        
    except ProcessingError as e:
        logger.error(f"❌ Processing error in get voices: {e.message}")
        raise HTTPException(status_code=500, detail=e.message)
        
    except Exception as e:
        logger.error(f"❌ Unexpected error in get voices: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to fetch voices: {str(e)}")